torch/distributions/multinomial.py

```diff
@@ -0,0 +1,91 @@
+    `probs` indexes over categories. All other dimensions index over batches.
+
+    Note that `total_count` need not be specified if only :meth:`log_prob` is
+    called (see example below)
+
+    -   :meth:`sample` requires a single shared `total_count` for all
+        parameters and samples.
+    -   :meth:`log_prob` allows different `total_count` for each parameter and
+        sample.
+
+    Example::
+
+        >>> m = Multinomial(10, torch.Tensor([ 1, 1, 1, 1]))
+        >>> x= m.sample()  # equal probability of 0, 1, 2, 3
+         1
+         3
+         3
+         3
+        [torch.FloatTensor of size 4]
+
+        >>> Multinomial(probs=torch.Tensor([1, 1, 1, 1])).log_prob(x)
+        -4.1338
+        [torch.FloatTensor of size 1]
+
+    Args:
+        total_count (int or Tensor or Variable): number of trials
+        probs (Tensor or Variable): event probabilities
+        logits (Tensor or Variable): event log probabilities
+    """
+    params = {'total_count': constraints.nonnegative_integer,
+              'probs': constraints.simplex, 'logits': constraints.real}
+
+    def __init__(self, total_count=1, probs=None, logits=None):
+        if isinstance(total_count, Number):
+            self.total_count = total_count
+        else:
+            if isinstance(total_count, Variable):
+                total_count = total_count.data
+            self.total_count = total_count.view(-1)[0]
+            if (total_count != self.total_count).any():
+                raise NotImplementedError('inhomogeneous total_count is not supported')
+        self._categorical = Categorical(probs=probs, logits=logits)
+        batch_shape = probs.size()[:-1] if probs is not None else logits.size()[:-1]
+        event_shape = probs.size()[-1:] if probs is not None else logits.size()[-1:]
+        super(Multinomial, self).__init__(batch_shape, event_shape)
+
+    @constraints.dependent_property
+    def support(self):
+        return constraints.integer_interval(0, self.total_count)
+
+    @property
+    def logits(self):
+        return self._categorical.logits
+
+    @property
+    def probs(self):
+        return self._categorical.probs
+
+    def sample(self, sample_shape=torch.Size()):
+        sample_shape = torch.Size(sample_shape)
+        samples = self._categorical.sample(torch.Size((self.total_count,)) + sample_shape)
+        # samples.shape is (total_count, sample_shape, batch_shape), need to change it to
+        # (sample_shape, batch_shape, total_count)
+        shifted_idx = list(range(samples.dim()))
+        shifted_idx.append(shifted_idx.pop(0))
+        samples = samples.permute(*shifted_idx)
+        counts = samples.new(self._extended_shape(sample_shape)).zero_()
+        counts.scatter_add_(-1, samples, torch.ones_like(samples))
+        return counts.type_as(self.probs)
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        logits, value = broadcast_all(self.probs.log(), value)
```

Comment:
The issue here was that we need to clone the logits because they are getting mutated in line 89 below. We can change that and resort to using `self.probs_or_logits` instead as part #4614.

Reply:
Can't we do `self.logits.clone()`? That's likely to be faster than `log()` + if someone gives you `logits` in the constructor we'll do `logits` -> `probs` (because `self.probs` is accessed) -> `logits`, which isn't a good idea